C error: invalid operand types for binary operation

### Describe the bug


Under some contexts, it seems the compiler doesn't notice that the '<' operator is not defined. 

If i have not defined how to compare (and not get a V error), the compiler proceeds to try to compile, failing

### Expected Behavior

I should get a V error saying that i didn't define the '<' operand

### Current Behavior

```v -g .``` output:
```
C:/Users/User/AppData/Local/Temp/v_0/02.8827659613261852381.tmp.c:499: warning: WINVER redefined
C:/Users/User/AppData/Local/Temp/v_0/../../../../../..C:\Program Files\v\vlib\builtin\builtin.c.v:68: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/User/AppData/Local/Temp/v_0/../../../../../..C:\Program Files\v\vlib\builtin\builtin_windows.c.v:288: warning: assignment from incompatible pointer type
C:/Users/User/AppData/Local/Temp/v_0/../../../../../..C:\Program Files\v\vlib\builtin\builtin_windows.c.v:288: warning: assignment makes pointer from integer without a cast
C:/Users/User/AppData/Local/Temp/v_0/../../../../../..C:\Program Files\v\vlib\builtin\builtin_windows.c.v:288: warning: cast between pointer and integer of different size
C:/Users/User/AppData/Local/Temp/v_0/../../../../../..C:\Users\User\Desktop\v_test\02\main.v:26: error: invalid operand types for binary operation
C:/Users/User/AppData/Local/Temp/v_0/../../../../../..C:\Users\User\Desktop\v_test\02\main.v:26: error: invalid operand types for binary operation
builder error:
==================
C error. This should never happen.
```

### Reproduction Steps

```
// The important function is percolateup, 
// the other functions are to make the situation that the function needs for it to be compiled
pub struct INode{
mut:
	heapindex u64
}

pub struct IHeap[T] {
mut:
	array []&T
	size  u64
}

pub fn create_iheap[T](capacity int) &IHeap[T] {
	mut h := &IHeap[T]{
		array: []&T{cap: capacity} 
		size: 0
	}
	h.array << &T(unsafe{nil})
	return h
}

fn (mut h IHeap[T]) percolateup(hl u64, mut element T) {
	mut hole := hl
	for hole > 1 && element < h.array[hole/2] {
		// ^ Notice that '<' was not defined
		println("didnt define '<' for the INode")
		hole /= 2
	}
}


fn main() {
	heap := create_iheap<INode>(3000000)
	println(heap)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.4 d8cf65d

### Environment details (OS name and version, etc.)

```
V full version: V 0.3.4 d8cf65d
OS: windows, Microsoft Windows 10 Pro v19045 64-bit
Processor: 8 cpus, 64bit, little endian,

getwd: C:\Users\User\Desktop\v_test\02
vexe: C:\Program Files\v\v.exe
vexe mtime: 2023-05-11 14:10:50

vroot: contains spaces, value: C:\Program Files\v
VMODULES: OK, value: C:\Users\User\.vmodules
VTMP: OK, value: C:\Users\User\AppData\Local\Temp\v_0

Git version: git version 2.40.1.windows.1
Git vroot status: weekly.2023.19-8-gd8cf65df
.git/config present: true

CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.
 cmd: cc --version
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -383,6 +383,10 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 			}
 		}
 		.gt, .lt, .ge, .le {
+			unwrapped_left_type := c.unwrap_generic(left_type)
+			left_sym = c.table.sym(unwrapped_left_type)
+			unwrapped_right_type := c.unwrap_generic(right_type)
+			right_sym = c.table.sym(unwrapped_right_type)
 			if left_sym.kind in [.array, .array_fixed] && right_sym.kind in [.array, .array_fixed] {
 				c.error('only `==` and `!=` are defined on arrays', node.pos)
 			} else if left_sym.kind == .struct_
@@ -392,8 +396,8 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 			} else if left_sym.kind == .struct_ && right_sym.kind == .struct_
 				&& node.op in [.eq, .lt] {
 				if !(left_sym.has_method(node.op.str()) && right_sym.has_method(node.op.str())) {
-					left_name := c.table.type_to_str(left_type)
-					right_name := c.table.type_to_str(right_type)
+					left_name := c.table.type_to_str(unwrapped_left_type)
+					right_name := c.table.type_to_str(unwrapped_right_type)
 					if left_name == right_name {
 						if !(node.op == .lt && c.pref.translated) {
 							// Allow `&Foo < &Foo` in translated code.

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -674,7 +674,8 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 				c.error('only `==`, `!=`, `|` and `&` are defined on `[flag]` tagged `enum`, use an explicit cast to `int` if needed',
 					node.pos)
 			}
-		} else if !c.pref.translated && !c.file.is_translated {
+		} else if !c.pref.translated && !c.file.is_translated && !left_type.has_flag(.generic)
+			&& !right_type.has_flag(.generic) {
 			// Regular enums
 			c.error('only `==` and `!=` are defined on `enum`, use an explicit cast to `int` if needed',
 				node.pos)

--- a/vlib/v/checker/tests/generics_undefined_operation_2.out
+++ b/vlib/v/checker/tests/generics_undefined_operation_2.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generics_undefined_operation_2.vv:23:18: error: undefined operation `&INode` < `&INode`
+   21 | fn (mut h IHeap[T]) percolateup(hl u64, mut element T) {
+   22 |     mut hole := hl
+   23 |     for hole > 1 && element < h.array[hole / 2] {
+      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   24 |         // ^ Notice that '<' was not defined
+   25 |         println("didnt define '<' for the INode")

--- a/vlib/v/checker/tests/generics_undefined_operation_2.vv
+++ b/vlib/v/checker/tests/generics_undefined_operation_2.vv
@@ -1,0 +1,33 @@
+pub struct INode {
+mut:
+	heapindex u64
+}
+
+pub struct IHeap[T] {
+mut:
+	array []&T
+	size  u64
+}
+
+pub fn create_iheap[T](capacity int) &IHeap[T] {
+	mut h := &IHeap[T]{
+		array: []&T{cap: capacity}
+		size: 0
+	}
+	h.array << &T(unsafe { nil })
+	return h
+}
+
+fn (mut h IHeap[T]) percolateup(hl u64, mut element T) {
+	mut hole := hl
+	for hole > 1 && element < h.array[hole / 2] {
+		// ^ Notice that '<' was not defined
+		println("didnt define '<' for the INode")
+		hole /= 2
+	}
+}
+
+fn main() {
+	heap := create_iheap[INode](3000000)
+	println(heap)
+}

